sklearn/preprocessing/_encoders.py

```diff
@@ -405,12 +405,14 @@ def transform(self, X):
             n_values = [len(cats) for cats in self.categories_]
 
         mask = X_mask.ravel()
-        n_values = np.array([0] + n_values)
-        feature_indices = np.cumsum(n_values)
+        feature_indices = np.cumsum([0] + n_values)
         indices = (X_int + feature_indices[:-1]).ravel()[mask]
-        indptr = X_mask.sum(axis=1).cumsum()
-        indptr = np.insert(indptr, 0, 0)
-        data = np.ones(n_samples * n_features)[mask]
+
+        indptr = np.empty(n_samples + 1, dtype=np.int)
+        indptr[0] = 0
+        np.sum(X_mask, axis=1, out=indptr[1:])
+        np.cumsum(indptr[1:], out=indptr[1:])
+        data = np.ones(indptr[-1])
```

Comment:
The last line is the only significant change, right? Or do the other changes also matter?

Reply:
They all matter. If I revert the last line, I get:

```
                      wall_time   cpu_time                                                                                                      
n_samples n_features                      
100000    50           0.575167   1.999209
          100          1.244284   2.790776
200000    50           1.424922   2.935314
          100          3.083203   4.711901
500000    50           3.762984   5.473710
          100          7.914155  10.137310
```